sdk/include/platform/sunburst/platform-i2c.hh

```diff
@@ -0,0 +1,385 @@
+	  StatusTransmitFull = 1 << 6,
+	  /// Target mode Receive FIFO is full
+	  StatusAcquiredFull = 1 << 7,
+	  /// Target mode Transmit FIFO is empty
+	  StatusTransmitEmpty = 1 << 8,
+	  /// Target mode Aquired FIFO is empty
+	  StatusAcquiredEmpty = 1 << 9,
+	  /**
+	   * A Host-Mode active transaction has been ended by the
+	   * HostNackHandlerTimeout mechanism. This bit is cleared when
+	   * Control.EnableHost is set by software to start a new transaction.
+	   */
+	  StatusHostDisabledNackTimeout = 1 << 10,
+	};
+
+	/// FormatData Register Fields
+	enum [[clang::flag_enum]] : uint32_t{
+	  /// Issue a START condition before transmitting BYTE.
+	  FormatDataStart = 1 << 8,
+	  /// Issue a STOP condition after this operation
+	  FormatDataStop = 1 << 9,
+	  /// Read BYTE bytes from I2C. (256 if BYTE==0)
+	  FormatDataReadBytes = 1 << 10,
+	  /**
+	   * Do not NACK the last byte read, let the read
+	   * operation continue
+	   */
+	  FormatDataReadCount = 1 << 11,
+	  /// Do not signal an exception if the current byte is not ACK’d
+	  FormatDataNakOk = 1 << 12,
+	};
+
+	/// FifoControl Register Fields
+	enum [[clang::flag_enum]] : uint32_t{
+	  /// Receive fifo reset. Write 1 to the register resets it. Read returns 0
+	  FifoControlReceiveReset = 1 << 0,
+	  /// Format fifo reset. Write 1 to the register resets it. Read returns 0
+	  FifoControlFormatReset = 1 << 1,
+	  /// Aquired FIFO reset. Write 1 to the register resets it. Read returns 0
+	  FifoControlAcquiredReset = 1 << 7,
+	  /// Transmit FIFO reset. Write 1 to the register resets it. Read returns 0
+	  FifoControlTransmitReset = 1 << 8,
+	};
+
+	/**
+	 * Specification times (Table 10) in nanoseconds for each bus mode.
+	 */
+	static constexpr uint16_t SpcThigh[]  = {4000u, 600u, 260u};
+	static constexpr uint16_t SpcTlow[]   = {4700u, 1300u, 150u};
+	static constexpr uint16_t SpcThdSta[] = {4000u, 600u, 260u};
+	static constexpr uint16_t SpcTsuSta[] = {4700u, 600u, 260u};
+	static constexpr uint16_t SpcThdDat[] = {4000u, 1u, 1u};
+	static constexpr uint16_t SpcTsuDat[] = {500u, 100u, 50u};
+	static constexpr uint16_t SpcTBuf[]   = {4700u, 1300u, 500u};
+	static constexpr uint16_t SpcTsuSto[] = {4000u, 600u, 260u};
+
+	/// Reset all of the fifos
+	void reset_fifos() volatile
+	{
+		fifoCtrl = (FifoControlReceiveReset | FifoControlFormatReset |
+		            FifoControlAcquiredReset | FifoControlTransmitReset);
+	}
+
+	/// Configure the I2C block to be in host mode
+	void set_host_mode() volatile
+	{
+		control = ControlEnableHost;
+	}
+
+	/**
+	 * Set the I2C timing parameters appropriately for the given bit rate.
+	 * Distilled from:
+	 * https://github.com/lowRISC/opentitan/blob/9ddf276c64e2974ed8e528e8b2feb00b977861de/hw/ip/i2c/doc/programmers_guide.md
+	 */
+	void set_speed(const uint32_t SpeedKhz) volatile
+	{
+		// We must round up the system clock frequency to lengthen intervals.
+		constexpr uint32_t SysclkKhz = (I2cDevClockHz + 999) / 1000;
```

Comment:
Is this doing the same round up thing as your helper?

Reply:
`round_up_divide` wasn't originally used because it truncates the result to a 16 bit value. This isn't a problem for the sonata's clock frequency, so I made the suggested change.

However, CI is now failing with the following due to this and the lower default clock in `compile_flags.txt`.

```
implicit conversion from ''int'' to ''const uint16_t'' (aka ''const unsigned short'') changes value from 500000 to 41248
```

Reply:
I was a bit confused about why the `round_up_divide` function returned a 16-bit result, since there’s no guarantee that this is adequate to represent the result. I’d be more comfortable with it returning a `uint32_t`. Or, at least, having an assert that the result is representable as 16 bits.

Reply:
We can probably bump the clock speed in compile flags. The combination of flags doesn’t have to match any real target. It just has to not break things that it checks.

Reply:
I've added an assert to the `round_up_divide` function